### PR TITLE
Avoid NaN values in torch.cdist backward for p<1

### LINKED ISSUE #36493: torch.cdist gradients are NAN for p<1 and very small differences in a given dimension (0<delta<~e-45)
## 🐛 Bug

When using fractional norm distances between a set of feature vectors (BS x Dim) and a set of class-centers (K x Dim) via torch.cdist(feat-vecs, class-vecs, p<1) as training objective, NANs occur in gradients when the difference between a value in dim D in a feature vector and a class vector is extremely small. There seem to be mysterious underflow issues in the backward pass for fractional norms, as the NANs do not occur when the difference is exactly 0.

## To Reproduce

1. Create sufficiently many random vectors to yield edge cases:

`import torch`
`bs = 40`
`k = 10000`
`dim = 1920`
`x = torch.randn(bs,dim).cuda()`
`w = torch.randn(k,dim).cuda()`
`x.requires_grad = True`
`w.requires_grad = True`

2. Calculate cdist for p=0.5, compute gradients and find NANs. By printing the indices in the dim dimension of the vectors and the specific values you get the idea that NANs occur for (nearly) identical values in the same dimension.

`y = torch.cdist(x,w, p=.5)`
`y.mean().backward()`
`x_nans = x.grad.ne(x.grad).nonzero()`
`w_nans = w.grad.ne(w.grad).nonzero()`

`print('Number of NANs found: ', len(x_nans), len(w_nans))`
`print('NANs found in dimensions:')`
`print([i[1].item()  for i in x_nans], [i[1].item()  for i in w_nans])`
`print('NANs found at values:')`
`print([x[x_nans[i][0],x_nans[i][1]].item() for i in range(len(x_nans))])`
`print([w[w_nans[i][0],w_nans[i][1]].item() for i in range(len(w_nans))])`

Example Output:

> 4 4
[1323, 1253, 1759, 1525] [1525, 1323, 1759, 1253]
[1.2044615745544434, 0.3567602336406708, -0.607083261013031, -0.5550623536109924]
[-0.5550623536109924, 1.2044615745544434, -0.607083261013031, 0.3567602336406708]

3. To see that it does not occur for truly identical values in a dim, you can simply replace x and y by zero-vecs and no NANs occur:
`x = torch.zeros(40,1920).float()`
`w = torch.zeros(10000,1920).float()`

4.  Alternatively, by using a computationally way slower alternative, we can avoid the NANs by adding a delta to the difference between two vectors, thereby pushing the min difference over the threshold delta. For me, delta = 1e-45 never yields NANs, while delta = 1e-46 always does. This workaround is a sad one though, because it increases training duration >>2-fold :(.

`delta = 1e-45`
`y = torch.stack([torch.norm(x[b]-w+delta, p=.5, dim=1) for b in range(x.size(0))])`

## Expected behavior

Efficient training using torch.cdist. Instead got a really nasty struggle to spot the source of those NANs. 

## Environment

 - PyTorch Version: 1.4
 - OS: Ubuntu 18.04
 - How you installed PyTorch:`conda`
 - Python version: 3.7
 - CUDA version: 10.1

## Additional Question:
Quite generally, is it possible to provide the efficient pairwise subtraction of two vectors that must be intrinsic to torch.cdist as a standalone function? I suspect, the inefficiency my workaround produces is caused by the pythonic list comprehension, while cdist seems to be implemented in C. Any suggestions for speedups I could gain myself?

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

### PR DESCRIPTION
**Pull request by kurtamohler**
Fixes #36493

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -6112,6 +6112,18 @@ def _test_euclidean_large_cdist(sizex, sizey=None):
         _test_cdist_for_size((1, 1), (S, 1))
         _test_euclidean_large_cdist((2000, 5))
 
+    # Ensure that cdist backward with p<1 does not produce NaNs
+    def test_cdist_grad_p_lt_1_no_nan(self, device):
+        for p in [0.99, 0.7, 0.5, 0.1, 0.01]:
+            x = torch.randn(1, 4, device=device)
+            y = x + torch.tensor([[1.0, 0.0, 1e-45, 1e-46]], device=device)
```
Comment (albanD) [inline test/test_autograd.py:6119]: Not sure if it's really useful to test 1e-45/1e-46.
If you just want to check the value at 0 that should be fine.

Note that if you actually use a small value that is not 0 (like 1e-6), you will most likely still get numerical issues (like we do for other similar ops: https://github.com/pytorch/pytorch/issues/43211).

File: test/test_autograd.py [hunk 2]
```diff
@@ -6112,6 +6112,18 @@ def _test_euclidean_large_cdist(sizex, sizey=None):
...
+            x.requires_grad = True
+            y.requires_grad = True
+            result = torch.cdist(x, y, p=0.5)
```
Comment (albanD) [inline test/test_autograd.py:6122]: I think you want p and not 0.5 here.

**Review by albanD — COMMENTED** [review]
Looks mostly good.
Added comments inline.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp [hunk 3]
```diff
@@ -104,7 +104,11 @@ struct Dist {
 
   // Special general pnorm derivative if p is less than two
   struct lttdist_calc {
-    static inline Vec backward(const Vec& diff, const scalar_t grad, const scalar_t dist, const Vec& p) { return dist == 0.0 ? Vec(0) : sign(diff) * diff.abs().pow(p - Vec(1)) * Vec(grad) / Vec(dist).pow(p - Vec(1)); }
+    static inline Vec backward(const Vec& diff, const scalar_t grad, const scalar_t dist, const Vec& p) {
+      Vec result = (dist == 0.0) ? Vec(0) : (sign(diff) * diff.abs().pow(p - Vec(1)) * Vec(grad) / Vec(dist).pow(p - Vec(1)));
+      result = Vec::blendv(result, Vec(0), (diff == Vec(0)) & (p < Vec(1)));
+      return result;
+    }
```
Comment (nikitaved) [inline aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: The gradients in here is defined for the
`cdist(x, y) = \sum_i (|x_i - y_i|^p)^{1/p}`. For values 0 < p < 1, this function is not a norm, and the underlying space is not a metric space. And it is not differentiable at zero, which means that the value of `result` could be made arbitrary large.
However, if for values 0 < p < 1 we ignore the p-root and consider:
`cdist(x, y) = \sum_i |x_i - y_i|^p`, now this function is a metric, so it does indeed measure distance.

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Another argument, the derivative limits from the left and right have huge values and of opposite sign. It means, that if we are to run gradient descent, it is very likely to zig-zag around zero without convergence.

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: The logic that we use for these is the following (similar to what we do for norm): The function is convex and admits a subgradient at 0. And we use the min-norm subgradient when multiple ones exists to ensure that it provides a descend direction. In this case 0.

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Case `p=0` is special, if we adopt the metric
`cdist(x, y)_p = \sum_i |x_i - y_i|^p`, then
`cdist(x, y)_0 = \sum_i {1 if |x_i - y_i| > 0}`

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Ah, OK, there is yet another definition of 0-"norm". More on the general case [here](https://en.wikipedia.org/wiki/Lp_space#When_0_%3C_p_%3C_1).
@albanD , your thoughts?

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: For `p=0` we should return 0 gradient all the time IIRC (piecewise constant function).
This is not properly special cased for cdist? There is a zdist above in this file but maybe this is not doing the right thing.

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: No, the function is actually concave for `0<p<1` :(

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: The problem is not with returning zero, but rather in what neighborhood to return it. So, the convention to return zero once in `eps` neighborhood limited by `1e-45`?

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: It's not always indeed but I would argue that this is fine as:
- This is true for the case where these is a single element as the abs() function is.
- When there are more elements, this remains below the whole function and touches it at the point of interest.
- We don't want to provide a non-zero gradient at zero for these functions as we would have to pick the sign arbitrarily. And we don't want to return nan as it is not user-friendly.

Can you think of another definition we can use here?

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: > The problem is not with returning zero, but rather in what neighborhood to return it. So, the convention to return zero once in eps neighborhood limited by 1e-45?

Ho, we don't have an answer to that I'm afraid.
Currently for all the norm-like functions, only `0` will have a custom value but values very close to 0 can generate inf/nan due to numerical precision issues (see the issue I liked above).

`1e-45` doesn't sound good (at least for float32) as you will get inf/nan way before reaching that.
The two options I can see are:
- Arbitrarily pick a range of value where we override the gradient (note that this range will be dtype dependant) but that might not solve all the problems and can be flaky.
- Inspect the computed gradient while knowing the formula and change inf/nan generated by the global formula if we detect that this is due to numerical precision (because the final value is close to 0)? But that would require some work to make this work properly for complex formulas like "norm" that can create inf/nan for many reasons beyond just the value being close to 0.

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Nothing against zero at zero, but the chance of this exact zero has probability zero. The function is not differentiable at zero, which means it is likely to get numerical issues at values close to zero. One possible resolution of this is by bounding the gradients from above by some max value, but then this bound depends on the inputs' dimension. Do you think this is feasible?

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Sorry, Alban, there is a delay with which our replies appear. Then what do you think about making a warning in the documentation of these norm-like functions with these specific values? Something like: "expect numerical issues around zero, consider bounding the gradients as a possible workaround", or something along these line?

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: > this exact zero has probability zero. 

Not in the amazing neural network world where half of your network contains `relu` that set things to 0 all the time haha
Also for distance functions, I wouldn't say that people do `dist(a, a)` with probability 0 (maybe not on purpose, but I have definitely already seen this).

> One possible resolution of this is by bounding the gradients from above by some max value

For p<1, this might be ok but mathematically, the gradient does go to infinity when you're close to 0 so I don't see why you shouldn't get such very large value? The analogy here for me is sqrt that will return larger and larger values as you get closer to 0 and it will eventually give you inf.
For p>1, this would actually be wrong as the gradient is, in theory, smoothly converging towards 0 at 0 and only numerical inaccuracies make it grow.

Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Ok, I am convinced :) Let the user decide then, he is smart enough to understand the issues. Sorry for the confusion!

Comment (albanD) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: :)
There is still an issue for p>1 for values very close to 0 though. But this is beyond the scope of this PR. But if you have ideas on how we can approach that, I am very interested!

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 4]
```diff
@@ -50,7 +50,9 @@ struct dists {
 
   // Special case backward when p is less than two
   struct lt_two {
-    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return dist == 0.0 ? 0 : sign(diff) * std::pow(std::abs(diff), p - 1) * grad / std::pow(dist, p - 1); }
+    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) {
+      return (dist == 0.0 || (diff == 0.0 && p < 1)) ? 0 : (sign(diff) * std::pow(std::abs(diff), p - 1) * grad / std::pow(dist, p - 1));
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/DistanceKernel.cu:54]:
```suggestion
      return (dist == 0.0 || (diff == 0.0 && p < 1)) ? 0 : (sign(diff) * std::pow(std::abs(diff), p - 1) * grad * std::pow(dist, 1 - p));
```
This resolves division by zero when `cdist` is close to zero, so should be more numerically stable.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/DistanceKernel.cu:54]: Or probably even more numerically stable approach:
`std::pow(std::abs(diff) / dist, p - 1) * grad`, or maybe we should try log-space, although it seems bad for zero.
`

Comment (kurtamohler) [reply aten/src/ATen/native/cuda/DistanceKernel.cu:54]: That change does not avoid the NaN by itself. It still needs the `(diff == 0.0 && p < 1)` check. Because of that, I feel like changing the formula in this way would be outside the scope of this PR, right?

File: test/test_autograd.py (hunk 1 above)
Comment (kurtamohler) [reply test/test_autograd.py]: Yeah, good point. I've removed 1e-45 and 1e-46.

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 4 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/DistanceKernel.cu:54]: It will be bc-breaking for sure, so I agree with you, better be kept for a different PR specifically targeting overall numerical stability.

File: aten/src/ATen/native/cpu/DistanceOpsKernel.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/cpu/DistanceOpsKernel.cpp:111]: Yeah, it seem we can rearrange things a bit to reduce the effect, but it is going to be bc-breaking I presume.

**Review by albanD — APPROVED** [review]
Looks good.
We can think of more "near to zero" optimization in a follow up PR if we want to.

**Comment by nikitaved** [comment]
@alban, we can modify the formula for different cases `p < 1`, `p > 1` to make it more numerically stable. It will be bc-breaking for sure.

**Comment by albanD** [comment]
Not all BC-breaking are a nogo. If we change the computed value to be closer to the correct thing, that's an OK thing to do.

The main thing is that we should make sure that it does not make other regions significantly worse.
